fix(scene-management): additional check to see if scene is already loaded

Myxcil opened the pull request:
…and can pass through

The previous implementation only checked if the scene to be loaded is matching the active scene. This caused problems in a multi-scene environment and non-active but already loaded scenes were loaded twice.

## Testing and Documentation
* No tests have been added.

Myxcil committed:
fix(scene-management): additional check to see if scene is already loaded and can pass through

The previous implementation only tested if the scene to be loaded is matching the active scene. This caused problems in a multi-scene environment and non-active but already loaded scenes were loaded twice.

Myxcil committed:
Merge pull request #1 from Myxcil/experimental/notruf

Additional check to see if the scene is already loaded.

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1418) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1418) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1418) it.</sub>

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -1496,7 +1496,18 @@ private void OnClientBeginSync(uint sceneEventId)
                 shouldPassThrough = true;
             }
-
+            else
```
This will cause conflicts with multiple versions of the same additive scene being loaded since their names will be identical and the only distinguishing difference would be the handle of the scene loaded.
However, I think I could make a change to the ValidateSceneBeforeLoading to bypass loading if you return false without completely exiting the synchronization process which would give you the same functionality I believe you are looking for?

Myxcil replied:
Hadn't encountered a situation where one has to load the same scene several times but your solution sounds much better - more control to the API user, yeah ;-)
I tried the ValidateScene/VerifySceneBeforeLoading delegate before, because it sounded exactly like what I was looking for, but as you said it stopped the sync process completely.

NoelStephensUnity commented:
I have a potential client side fix that I think should be implemented anyway, but you might be able to handle the issue you are having by assigning a server-side ValidateSceneBeforeLoading callback that would exclude the scenes that you are expecting to be inactive but loaded on the client side from being included in the synchronization response from the server.  This happens in NetworkSceneManager.SynchronizeNetworkObjects in these two locations:
[Location-1](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/45a0c8b8659bd7a9c738388dfd479ef2a49ede7f/com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs#L1415)
[Location-2](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/45a0c8b8659bd7a9c738388dfd479ef2a49ede7f/com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs#L1422)
_The idea is that the server just excludes those scenes from being loaded on the client side._

However. if this is a scenario where it is not always true (_i.e. some clients have inactive scenes already loaded and some clients don't have the scene loaded_) then the client-side suggested adjustment to ValidateSceneBeforeLoading would be able to provide you with the ability to handle that scenario too.

Let me know if the server side fix works for you?

Myxcil committed:
Merge branch 'Unity-Technologies:develop' into develop

Myxcil committed:
Merge branch 'Unity-Technologies:develop' into experimental/notruf

Myxcil committed:
- reverted https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1418

Myxcil committed:
Merge branch 'develop' into experimental/notruf

Myxcil commented:
> Let me know if the server side fix works for you?
Removed my "fix" and added a VerifySceneBeforeLoading on the server side, which returns false for all scenes but the one containing the NetworkObjects. Looking good so far. Thanks for your help! 🤗

NoelStephensUnity commented:
@Myxcil 
Glad to hear that worked for you.  If you have any other suggested fixes please feel free to submit another PR.  If you run into any snags with this alternate solution feel free to file a github issue.  
Cheers!  :)
